packages/flutter/lib/src/material/navigation_bar.dart

```diff
@@ -462,38 +462,36 @@ class _NavigationDestinationBuilder extends StatelessWidget {
     final _NavigationDestinationInfo info = _NavigationDestinationInfo.of(context);
     final NavigationBarThemeData navigationBarTheme = NavigationBarTheme.of(context);
     final NavigationBarThemeData defaults = _defaultsFor(context);
+    final ValueNotifier<Offset> iconOffset = ValueNotifier<Offset>(Offset.zero);
 
-    final bool selected = info.selectedIndex == info.index;
-    final double labelPadding;
-    switch (info.labelBehavior) {
-      case NavigationDestinationLabelBehavior.alwaysShow:
-        labelPadding = 8;
-        break;
-      case NavigationDestinationLabelBehavior.onlyShowSelected:
-        labelPadding = selected ? 8 : 0;
-        break;
-      case NavigationDestinationLabelBehavior.alwaysHide:
-        labelPadding = 0;
-        break;
-    }
     return _NavigationBarDestinationSemantics(
       child: _NavigationBarDestinationTooltip(
         message: tooltip ?? label,
-        child: _IndicatorInkWell(
-          key: UniqueKey(),
-          labelPadding: labelPadding,
-          customBorder: navigationBarTheme.indicatorShape ?? defaults.indicatorShape,
-          onTap: info.onTap,
-          child: Row(
-            children: <Widget>[
-              Expanded(
-                child: _NavigationBarDestinationLayout(
-                  icon: buildIcon(context),
-                  label: buildLabel(context),
-                ),
+        child: ValueListenableBuilder<Offset>(
+          valueListenable: iconOffset,
+          builder: (BuildContext context, Offset offset, Widget? child) {
+            return _IndicatorInkWell(
+              key: UniqueKey(),
+              iconOffset: offset,
+              customBorder: navigationBarTheme.indicatorShape ?? defaults.indicatorShape,
+              onTap: info.onTap,
+              child: Row(
+                children: <Widget>[
+                  Expanded(
+                    child: _NavigationBarDestinationLayout(
+                      icon: buildIcon(context),
+                      label: buildLabel(context),
+                      iconOffsetChanged :(Offset value) {
+                        WidgetsBinding.instance.addPostFrameCallback((Duration timeStamp) {
```

Comment:
Fixing this up a frame late doesn't seem like a good solution. If the InkWell will not need the icon's position until after layout has happened then couldn't we look up the icon's position via a GlobalKey?

Reply:
Thanks for the suggestion. I was hesitant to use a GlobalKey but given the requirement here, I think it's worth it.

Updating the with GlobalKey to check for label size, as it is different between M2 and M3.